polkit/validate/validate.go

```diff
@@ -0,0 +1,297 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package validate
+
+import (
+	"encoding/xml"
+	"fmt"
+	"io"
+	"strings"
+)
+
+type Element struct {
+	CharData      string     `xml:",chardata"`
+	ExtraAttrs    []xml.Attr `xml:",any,attr"`
+	ExtraChildren []xml.Name `xml:",any"`
+}
+
+type policyConfig struct {
+	XMLName xml.Name
+	Element
+
+	Vendor    []Element `xml:"vendor"`
+	VendorURL []Element `xml:"vendor_url"`
+	IconName  []Element `xml:"icon_name"`
+
+	Actions []action `xml:"action"`
+}
+
+type action struct {
+	Element
+
+	ID string `xml:"id,attr"`
+
+	Vendor    []Element `xml:"vendor"`
+	VendorURL []Element `xml:"vendor_url"`
+	IconName  []Element `xml:"icon_name"`
+
+	Description []message  `xml:"description"`
+	Message     []message  `xml:"message"`
+	Defaults    []defaults `xml:"defaults"`
+	Annotate    []annotate `xml:"annotate"`
+}
+
+type message struct {
+	Element
+
+	GettextDomain string `xml:"gettext-domain,attr"`
+	Language      string `xml:"lang,attr"` // to match xml:lang
+}
+
+type defaults struct {
+	Element
+
+	AllowAny      []Element `xml:"allow_any"`
+	AllowInactive []Element `xml:"allow_inactive"`
+	AllowActive   []Element `xml:"allow_active"`
+}
+
+type annotate struct {
+	Element
+
+	Key string `xml:"key,attr"`
+}
+
+func ValidatePolicy(r io.Reader) (actionIDs []string, err error) {
+	decoder := xml.NewDecoder(r)
```

Comment:
shouldn't we use Strict mode? if not we should probably have a comment here explaining why not

Reply:
The documentation says strict mode is the default, and the implementation agrees:

https://github.com/golang/go/blob/296ddf2a936a30866303a64d49bc0e3e034730a8/src/encoding/xml/xml.go#L226-L238

Reply:
silly me